### PR TITLE
Make toolchain registration optional

### PR DESCRIPTION
**Pull request by layus**

### CODE REVIEW

File: nixpkgs/nixpkgs.bzl [hunk 1]
```diff
@@ -729,6 +731,11 @@ def nixpkgs_cc_configure(
         fail_not_supported = fail_not_supported,
     )
 
+    # Stop before generating the toolchain. The only point of not registering it
+    # is because it is not used.
+    if (not register):
+        return
+
```
Comment (aherrmann) [inline nixpkgs/nixpkgs.bzl:738]: Could you move this behind `_nixpkgs_cc_toolchain` before `maybe(native.bind, ...)`?

Creating the `toolchain`s when not registering them may seem redundant, but this is relevant for `bzlmod` compatibility, see [rules_sh#18](https://github.com/tweag/rules_sh/pull/18/files#diff-803e299a5866fd34cb4894026eab5103818f5d96ed8707af0af3ccecbc99e409R353-R354). In that case we still want to use the generated `toolchain`s, we just cannot use `native.register_toolchain` and instead have to use the new [`module` mechanism](https://github.com/tweag/rules_sh/pull/18/files#diff-6136fc12446089c3db7360e923203dd114b6a1466252e71667c6791c20fe6bdcR5).

Note, a `toolchain` rule that is not registered will not participate in toolchain resolution. So, this is still compatible with a use-case like [build-settings influencing toolchain resolution](https://github.com/aherrmann/bazel-toolchain-version-select).

Comment (layus) [reply nixpkgs/nixpkgs.bzl]: Done, thanks :-).

**Review by aherrmann — APPROVED** [review]
Thank you!

**Comment by aherrmann** [comment]
@Mergifyio rebase

**Comment by aherrmann** [comment]
@layus CI fails
```
FAIL: //docs:check-readme (see /home/runner/.cache/bazel/_bazel_runner/f390b00a27990596a8f9790e9a873450/execroot/io_tweag_rules_nixpkgs/bazel-out/k8-fastbuild/testlogs/docs/check-readme/test.log)
INFO: From Testing //docs:check-readme:
==================== Test output for //docs:check-readme:
The project README is not up-to-date.
Please update it using the following command.

  bazel run //docs:update-readme

================================================================================
```

**Pull request merged**